extensions/amp-ad/0.1/amp-ad-xorigin-iframe-handler.js

```diff
@@ -125,34 +124,58 @@ export class AmpAdXOriginIframeHandler {
       this.sendEmbedInfo_(this.baseInstance_.isInViewport());
     }));
 
+    // Iframe.onload normally called by the Ad after full load.
+    const iframeLoadPromise = loadPromise(this.iframe).then(() => {
+      // Wait just a little to allow messages to arrive.
+      return timer.promise(10);
+    });
     if (this.baseInstance_.emitLifecycleEvent) {
       // Only set up a load listener if we know that we can send lifecycle
       // messages.
-      loadPromise(this.iframe).then(() => {
+      iframeLoadPromise.then(() => {
         this.baseInstance_.emitLifecycleEvent('xDomIframeLoaded');
       });
     }
 
-    // Install API that listens to ad response
-    if (this.baseInstance_.config
-        && this.baseInstance_.config.renderStartImplemented) {
-      // If support render-start, create a race between render-start no-content
-      this.adResponsePromise_ = listenForOncePromise(this.iframe,
-        ['render-start', 'no-content'], true).then(info => {
-          const data = info.data;
-          if (data.type == 'render-start') {
-            this.renderStart_(info);
-          } else {
-            this.noContent_();
-          }
-        });
+    // Calculate render-start and no-content signals.
+    let renderStartResolve;
+    const renderStartPromise = new Promise(resolve => {
+      renderStartResolve = resolve;
+    });
+    let noContentResolve;
+    const noContentPromise = new Promise(resolve => {
+      noContentResolve = resolve;
+    });
+    if (this.baseInstance_.config &&
+            this.baseInstance_.config.renderStartImplemented) {
+      // When `render-start` is supported, these signals are mutually
+      // exclusive. Whichever arrives first wins.
+      listenForOncePromise(this.iframe,
+          ['render-start', 'no-content'], true).then(info => {
+            const data = info.data;
+            if (data.type == 'render-start') {
```

Comment:
Couple of things here:
 1. They are mutually exclusive. The first one wins, but they do have significant consequences to DOM/UX.
 2. The second branch (the one that doesn't support renderStart) is going away eventually.